keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -1172,6 +1210,183 @@ back to aggregating container requests.
+with resize policy, and how containers with specified resources remain unaffected by
+pod-level resizes.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name:     pod-level-resources
+spec:
+  resources:
+    requests:
+      cpu: 100m
+      memory: 100Mi
+    limits:
+      cpu: 200m
+      memory: 200Mi
+  containers:
+    - name: c1
+      image: registry.k8s.io/pause:latest
+      resizePolicy:
+        - resourceName: "cpu"
+          restartPolicy: "NotRequired"
+        - resourceName: "memory"
+          restartPolicy: "RestartRequired"
+    - name: c2
+      image: registry.k8s.io/pause:latest
+      resources:
+        requests:
+          cpu: 50m
+          memory: 50Mi
+        limits:
+          cpu: 100m
+          memory: 100Mi
+      resizePolicy:
+        - resourceName: "cpu"
+          restartPolicy: "NotRequired"
+        - resourceName: "memory"
+          restartPolicy: "RestartRequired"
+```
+
+In this example:
+* CPU resizes: Neither container requires a restart for CPU resizes, and therefore CPU resizes at neither the container nor pod level will trigger any restarts.
+* Container c1 (inherited memory limit): c1 does not define any container level
+  resources, so the effective memory limit of the container is determined by the
+  pod-level limit. When the pod's limit is resized, c1's effective memory limit
+  changes. Because c1's memory resizePolicy is RestartRequired, a resize of the
+  pod-level memory limit will trigger a restart of container c1.
+* Container c2 (specified memory limit): c2 does define container-level resources,
+  so the effective memory limit of c2 is the container level limit. Therefore, a
+  resize of the pod-level memory limit doesn't change the effective container limit,
+  so the c2 is not restarted when the pod-level memory limit is resized.
+
+##### Implementation Details
+
+###### Allocating Pod-level Resources
+Allocation of pod-level resources will work the same as container-level resources. The allocated resources checkpoint will be extended to include pod-level resources, and the pod object will be updated with the allocated resources in the pod sync loop.
+
+###### Actuating Pod-level Resource Resize
+The mechanism for actuating pod-level resize remains largely unchanged from the
+existing container-level resize process.  When pod-level resource configurations are
+applied, the system handles the resize in a similar manner as it does for
+container-level resources. This includes extending the existing logic to incorporate
+directly configured pod-level resource settings.
+
+The same ordering rules for pod and container resource resizing will be applied for each
+resource as needed:
+1. Increase pod-level cgroup (if needed)
+2. Decrease container resources
+3. Decrease pod-level cgroup (if needed)
+4. Increase container resources
+
+###### Tracking Actual Pod-level Resources
+To accurately track actual pod-level resources during in-place pod resizing, several
```

Comment:
@tallclair Given the discussion on how NRI plugins or systemd can mutate the resources (e.g. rounding), what happens when:

* User specifies a pod with 2 containers, each using `5m` cpu and the pod is `10m`
* An NRI plugin mutates the containers and rounds them up to `10m` each

Are we smart enough to increase the pod?

Reply:
Hmm, we don't today, but we could. We are reading the actual values from the runtime, so we could compute the the pod-level cgroups based on the sum of those instead of the allocated resources (or whichever is larger)

Reply:
Hmm... but if NRI plugins change the value to be completely different, that'd just conflicts with how kubelet manages the cgroups. We can simply grab the values from the and assume those are the ones we want. 

@samuelkarp

Reply:
Agreed with Yuju above. 

I am expecting that with those new efforts on the resource management we are doing and plan to do, the users eventually limit their NRI usages largely.

Reply:
We can continue discussing this, but this isn't a blocker here.